### PR TITLE
API resource pruning in pruner control loop

### PR DESCRIPTION
**Pull request by damemi**
Rework of https://github.com/openshift/library-go/pull/162 to prune API resources in the pruner control loop instead of the pruner pod, by adding a separate workqueueand running each in their own routines

### CODE REVIEW

File: pkg/operator/staticpod/controller/prune/prune_controller.go [hunk 1]
```diff
@@ -71,7 +73,8 @@ func NewPruneController(
 		kubeClient:           kubeClient,
 		eventRecorder:        eventRecorder,
 
-		queue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "PruneController"),
+		diskQueue:        workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "PruneDiskController"),
+		apiQueue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "PruneAPIController"),
```
[the remark this answers is not in the document]
Comment (damemi) [reply pkg/operator/staticpod/controller/prune/prune_controller.go]: We can do that too, I was just basing this off your comment here: https://github.com/openshift/library-go/pull/162#issuecomment-452472894

Though I do like the separate backoff/retry mechanics for 2 queues it's not like it makes a huge difference

File: pkg/operator/staticpod/controller/prune/prune_controller.go [hunk 2]
```diff
@@ -131,9 +136,44 @@ func (c *PruneController) pruneRevisionHistory(operatorStatus *operatorv1.Static
...
+func (c *PruneController) pruneAPIResources(operatorStatus *operatorv1.StaticPodOperatorStatus, excludedIDs []int, maxEligibleRevisionID int) error {
+	protectedIDs := sets.NewInt(excludedIDs...)
+	statusConfigMaps, err := c.kubeClient.CoreV1().ConfigMaps(c.targetNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, cm := range statusConfigMaps.Items {
+		revision, err := strconv.Atoi(cm.Data["revision"])
+		if err != nil {
+			return err
```
Comment (mfojtik) [inline pkg/operator/staticpod/controller/prune/prune_controller.go]: nit: add some context to this error

File: pkg/operator/staticpod/controller/prune/prune_controller.go [hunk 3]
```diff
@@ -131,9 +136,44 @@ func (c *PruneController) pruneRevisionHistory(operatorStatus *operatorv1.Static
...
+		}
+
+		if protected := protectedIDs.Has(revision); protected {
```
Comment (mfojtik) [inline pkg/operator/staticpod/controller/prune/prune_controller.go]: nit: you don't need the `protected` var

File: pkg/operator/staticpod/controller/prune/prune_controller.go [hunk 4]
```diff
@@ -131,9 +136,44 @@ func (c *PruneController) pruneRevisionHistory(operatorStatus *operatorv1.Static
...
+			continue
+		}
+		if revision > maxEligibleRevisionID {
+			continue
+		}
+		err = c.kubeClient.CoreV1().ConfigMaps(c.targetNamespace).Delete(cm.Name, &metav1.DeleteOptions{})
```
Comment (mfojtik) [inline pkg/operator/staticpod/controller/prune/prune_controller.go]: nit: `if err := ...`

File: pkg/operator/staticpod/controller/prune/prune_controller.go [hunk 5]
```diff
@@ -131,9 +133,42 @@ func (c *PruneController) pruneRevisionHistory(operatorStatus *operatorv1.Static
...
+	}
+	return excludedIDs, nil
+}
+
+func (c *PruneController) pruneDiskResources(operatorStatus *operatorv1.StaticPodOperatorStatus, excludedIDs []int, maxEligibleRevisionID int) error {
 	// Run pruning pod on each node and pin it to that node
 	for _, nodeStatus := range operatorStatus.NodeStatuses {
-		if err := c.ensurePrunePod(nodeStatus.NodeName, excludedIDs[len(excludedIDs)-1], excludedIDs, nodeStatus.TargetRevision); err != nil {
+		if err := c.ensurePrunePod(nodeStatus.NodeName, maxEligibleRevisionID, excludedIDs, nodeStatus.TargetRevision); err != nil {
+			return err
```
Comment (deads2k) [inline pkg/operator/staticpod/controller/prune/prune_controller.go]: as a followup, try to prune on all nodes and aggregate the errors.

**Pull request merged**